internal/api/token_oidc.go

```diff
@@ -0,0 +1,218 @@
+
+	default:
+		log.WithField("issuer", p.Issuer).WithField("client_id", p.ClientID).Warn("Use of POST /token with arbitrary issuer and client_id is deprecated for security reasons. Please switch to using the API with provider only!")
+
+		allowed := false
+		for _, allowedIssuer := range config.External.AllowedIdTokenIssuers {
+			if p.Issuer == allowedIssuer {
+				allowed = true
+				providerType = allowedIssuer
+				acceptableClientIDs = []string{p.ClientID}
+				issuer = allowedIssuer
+				break
+			}
+		}
+
+		if !allowed {
+			return nil, "", nil, badRequestError(fmt.Sprintf("Custom OIDC provider %q not allowed", p.Issuer))
+		}
+	}
+
+	if cfg != nil && !cfg.Enabled {
+		return nil, "", nil, badRequestError(fmt.Sprintf("Provider (issuer %q) is not enabled", issuer))
+	}
+
+	oidcProvider, err := oidc.NewProvider(ctx, issuer)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	return oidcProvider, providerType, acceptableClientIDs, nil
+}
+
+// IdTokenGrant implements the id_token grant type flow
+func (a *API) IdTokenGrant(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	log := observability.GetLogEntry(r)
+
+	db := a.db.WithContext(ctx)
+	config := a.config
+
+	params := &IdTokenGrantParams{}
+
+	body, err := getBodyBytes(r)
+	if err != nil {
+		return badRequestError("Could not read body").WithInternalError(err)
+	}
+
+	if err := json.Unmarshal(body, params); err != nil {
+		return badRequestError("Could not read id token grant params: %v", err)
+	}
+
+	if params.IdToken == "" {
+		return oauthError("invalid request", "id_token required")
+	}
+
+	if params.Provider == "" && (params.ClientID == "" || params.Issuer == "") {
+		return oauthError("invalid request", "provider or client_id and issuer required")
+	}
+
+	oidcProvider, providerType, acceptableClientIDs, err := params.getProvider(ctx, config, r)
+	if err != nil {
+		return err
+	}
+
+	idToken, userData, err := provider.ParseIDToken(ctx, oidcProvider, nil, params.IdToken, provider.ParseIDTokenOptions{
+		SkipAccessTokenCheck: params.AccessToken == "",
+		AccessToken:          params.AccessToken,
+	})
+	if err != nil {
+		return oauthError("invalid request", "Bad ID token").WithInternalError(err)
+	}
+
+	if idToken.Subject == "" {
+		return oauthError("invalid request", "Missing sub claim in id_token")
+	}
+
+	correctAudience := false
+	for _, clientID := range acceptableClientIDs {
+		if clientID == "" {
+			continue
+		}
+
+		for _, aud := range idToken.Audience {
+			if aud == clientID {
+				correctAudience = true
+				break
+			}
+		}
+
+		if correctAudience {
+			break
+		}
+	}
+
+	if !correctAudience {
+		return oauthError("invalid request", "Unacceptable audience in id_token")
+	}
+
+	tokenHasNonce := idToken.Nonce != ""
+	paramsHasNonce := params.Nonce != ""
+
+	if tokenHasNonce != paramsHasNonce {
+		return oauthError("invalid request", "Passed nonce and nonce in id_token should either both exist or not.")
```

Comment:
Do we need to check at all if `nonce` exists if `access_token` was sent as a parameter? 

When I was taking a look at how Firebase does Google login, Firebase Auth seem to accept access token and id token, while not accepting nonce, so was just wondering.